contracts/optimism/TokenRateOracle.sol

```diff
@@ -93,17 +111,78 @@ contract TokenRateOracle is CrossDomainEnabled, ITokenRateOracle, Versioned {
         TOKEN_RATE_OUTDATED_DELAY = tokenRateOutdatedDelay_;
         MAX_ALLOWED_L2_TO_L1_CLOCK_LAG = maxAllowedL2ToL1ClockLag_;
         MAX_ALLOWED_TOKEN_RATE_DEVIATION_PER_DAY = maxAllowedTokenRateDeviationPerDay_;
+        MAX_DELTA_TIME_TO_PAUSE_TOKEN_RATE_UPDATES = maxDeltaTimeToPausetokenRateUpdates_;
     }
 
-    function initialize(uint256 tokenRate_, uint256 rateL1Timestamp_) external {
+    /// @notice Initializes the contract from scratch.
+    /// @param admin_ Address of the account to grant the DEFAULT_ADMIN_ROLE
+    /// @param tokenRate_ wstETH/stETH token rate, uses 10**DECIMALS precision.
+    /// @param rateUpdateL1Timestamp_ L1 time when rate was updated on L1 side.
+    function initialize(address admin_, uint256 tokenRate_, uint256 rateUpdateL1Timestamp_) external {
+        if (admin_ == address(0)) {
+            revert ErrorZeroAddressAdmin();
+        }
         if (tokenRate_ < MIN_ALLOWED_TOKEN_RATE || tokenRate_ > MAX_ALLOWED_TOKEN_RATE) {
             revert ErrorTokenRateInitializationIsOutOfAllowedRange(tokenRate_);
         }
-        if (rateL1Timestamp_ > block.timestamp + MAX_ALLOWED_L2_TO_L1_CLOCK_LAG) {
-            revert ErrorL1TimestampInitializationIsOutOfAllowedRange(rateL1Timestamp_);
+        if (rateUpdateL1Timestamp_ > block.timestamp + MAX_ALLOWED_L2_TO_L1_CLOCK_LAG) {
+            revert ErrorL1TimestampInitializationIsOutOfAllowedRange(rateUpdateL1Timestamp_);
         }
         _initializeContractVersionTo(1);
-        _setTokenRateAndTimestamps(uint128(tokenRate_), uint64(rateL1Timestamp_));
+        _grantRole(DEFAULT_ADMIN_ROLE, admin_);
+        _addTokenRate(uint128(tokenRate_), uint64(rateUpdateL1Timestamp_), uint64(block.timestamp));
+    }
+
+    /// @notice Pauses token rate updates and sets the old rate provided by tokenRateIndex_.
+    ///         Should be called by DAO or emergency breaks only.
+    /// @param tokenRateIndex_ The index of the token rate that applies after the pause.
+    ///        Token Rate can't be received older then MAX_DELTA_TIME_TO_PAUSE_TOKEN_RATE_UPDATES
+    ///        except only if the passed index is the latest one.
+    function pauseTokenRateUpdates(uint256 tokenRateIndex_) external onlyRole(RATE_UPDATE_DISABLER_ROLE) {
+        if (_isPaused()) {
+            revert ErrorAlreadyPaused();
+        }
+        TokenRateData memory tokenRateData = _getTokenRateByIndex(tokenRateIndex_);
+        if (tokenRateIndex_ != _getStorageTokenRates().length - 1 &&
```

Comment:
may add a check that the length is not zero, hmm (nitpick)

Reply:
length can't be zero after initialization